### DISCUSSION TITLE
Injecting User model into FormRequest

### DISCUSSION THREAD

**Discussion by kimitaka** [laravel/framework#42443] [Q&A]
Hello,

I have a question about dependency injection in Laravel.

I want to access a `User` model instance that indicates currently manipulating user data in `FormRequest` to reduce controller size.
For example, to check an email is unique except for the current user; I wrote like this.

#### `FormRequest` implementation
```php
class UpdateEmailRequest extends FormRequest
{
    public function rules(User $user): array
    {
        return [
            'email' => ['required', 'email', Rule::unique('users', 'email')->ignore($user->id)],
        ];
    }
}
```

And I use this as a type hinting in a controller.

#### `Controller` implementation
```php
class UserController extends Controller
    public function update(UpdateEmailRequest $request)
    {
        // codes to update email
    }
```

I supposed Laravel's DI injects a currently manipulating `User` model object into `UpdateEmailRequest` by a relation between the controller and it.
But when I tried them, injected `$user` instance was a brand new `User` one.

I escaped this problem by accessing a request via route definition like these.

#### Route definition
```php
Route::put('/users/{user}', 'update');
```

#### In `UpdateEmailRequest` class
```php
$this->route('user')
```

However, I think this workaround is not a good idea because this means `UpdateEmailRequest` depends on a route definition.
Do you have any ideas for getting a `User` instance and keeping sparse connections among classes?
I beg your help, thank you.

**Comment by shaedrich** [comment]
When you want to get the currently authenticated user making this request inside the form request class, you can do so via `$request->user()`

**Reply by kimitaka** [reply]
Thank you for answering my question. However, I want to retrieve a user resource indicated by a URI like `/users/{id}`, not the authenticated user. I now understand how to obtain a user instance by manually fetching the request parameters. But by doing this, I cannot take full advantage of DI.
My question was posted a year ago, and I'm unfamiliar with the latest version of Laravel. I'll give it a try later.

**Reply by shaedrich** [reply]
Ah, then I misread "I think this workaround is not a good idea because this means UpdateEmailRequest depends on a route definition.". Thanks for clarifying that.
